### PR TITLE
[dask] merge local_predict tests into other tests (fixes #3833)

### LINKED ISSUE #3833: Merge test_<est>_local_predict and test_<est> tests for Dask module
Refer to https://github.com/microsoft/LightGBM/pull/3708#discussion_r562147641 for more details.

### PR DESCRIPTION
**Pull request by ShrillShrestha**
…#3833

### CODE REVIEW

File: tests/python_package_test/test_dask.py [hunk 1]
```diff
@@ -133,6 +133,11 @@ def test_classifier(output, centers, client, listen_port):
         centers=centers
     )
 
+    X_1, y_1, w_1, dX_1, dy_1, dw_1 = _create_data(
+        objective='classification',
+        output='array'
+    )
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:139]: Why was this necessary? You should just use the `dask_classifier` defined below this. With this change, you'd only be doing the local predict on arrays each time, but we want to test on all of DataFrame, Array, and sparse matrix.

File: tests/python_package_test/test_dask.py [hunk 2]
```diff
@@ -291,6 +278,8 @@ def test_regressor(output, client, listen_port):
     if output != 'dataframe':
         s1 = r2_score(dy, p1)
     p1 = p1.compute()
+    p2_local_predict = dask_regressor.to_local().predict(X)
+    s2_local_predict = dask_regressor.to_local().score(X, y)
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:282]:
```suggestion
    p2_local = dask_regressor.to_local().predict(X)
    s2_local = dask_regressor.to_local().score(X, y)
```

Please simplify these names a bit (and elsewhere in the PR) by removing the "_predict" suffix

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your interest in LightGBM and your contribution! Can you please explain why you needed to add an additional `_create_data()` call and an additional `.fit()` in the classifier test? That should not be necessary.

File: tests/python_package_test/test_dask.py (hunk 1 above)
Comment (ShrillShrestha) [reply tests/python_package_test/test_dask.py:139]: Initially, I tried to reference the prediction inside `test_classifier` with `p1_local=dask_classifier.to_local().predict(dX)` for `p1` of `test_classifier_local_predict`, but this line gave me two errors saying `ValueError: Expected 2D array, got scalar array instead` while running the tests.
I thought it was because of the difference we have in the `output` params in 
```
X, y, w, dX, dy, dw = _create_data(
        objective='classification',
        output=output,
        centers=centers
    )
```
and 
```
X, y, w, dX, dy, dw = _create_data(
        objective='classification',
        output='array'
    )
```
in `test_classifier` and `test_classifier_local_predict` accordingly. Is there any way to resolve it?

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:139]: `dask_classifier.to_local()` produces a `lightgbm.sklearn.LGBMClassifier`. Pass it `X` (a non-Dask data structure) instead of `dX`(a Dask data structure) and it should work without needing to have this extra `_create_data()`

Comment (ShrillShrestha) [reply tests/python_package_test/test_dask.py]: Thanks, it worked. I also renamed the variables. you can tell me if I need any other changes?

File: tests/python_package_test/test_dask.py [hunk 3]
```diff
@@ -291,6 +265,8 @@ def test_regressor(output, client, listen_port):
     if output != 'dataframe':
         s1 = r2_score(dy, p1)
     p1 = p1.compute()
+    p2_local = dask_regressor.to_local().predict(X)
+    s2_local = dask_regressor.to_local().score(X, y)
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:269]:
```suggestion
    p1_local = dask_regressor.to_local().predict(X)
    s1_local = dask_regressor.to_local().score(X, y)
```

Please call these `p1_local` and `p2_local`. We have a convention in these tests right now that `p1` references predictions from a model trained with `DaskLGBMRegressor` and the `p2` references things from `lightgbm.sklearn.LGBMRegressor`. I know that's confusing and we should probably change it, but for now this test should be consistent with the existing code.

File: tests/python_package_test/test_dask.py [hunk 4]
```diff
@@ -142,9 +142,11 @@ def test_classifier(output, centers, client, listen_port):
         local_listen_port=listen_port,
         **params
     )
+
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:145]:
```suggestion
```

please revert this unrelated whitespace change

File: tests/python_package_test/test_dask.py [hunk 5]
```diff
@@ -153,12 +155,13 @@ def test_classifier(output, centers, client, listen_port):
     p2 = local_classifier.predict(X)
     p2_proba = local_classifier.predict_proba(X)
     s2 = local_classifier.score(X, y)
-
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:156]: please revert this unrelated whitespace change and add this line back

File: tests/python_package_test/test_dask.py [hunk 6]
```diff
@@ -428,6 +382,7 @@ def test_ranker(output, client, listen_port, group):
     dask_ranker = dask_ranker.fit(dX, dy, sample_weight=dw, group=dg, client=client)
     rnkvec_dask = dask_ranker.predict(dX)
     rnkvec_dask = rnkvec_dask.compute()
+    rnkvec_local_1 = dask_ranker.to_local().predict(X)
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:385]:
```suggestion
    rnkvec_dask_local = dask_ranker.to_local().predict(X)
```

Please change this name so it's clearer how it differs from `rnkvec_local`

File: tests/python_package_test/test_dask.py [hunk 7]
```diff
@@ -438,35 +393,7 @@ def test_ranker(output, client, listen_port, group):
...
-        min_child_samples=1
-    )
-    dask_ranker = dask_ranker.fit(dX, dy, group=dg, client=client)
-    rnkvec_dask = dask_ranker.predict(dX)
-    rnkvec_dask = rnkvec_dask.compute()
-    rnkvec_local = dask_ranker.to_local().predict(X)
-
-    # distributed and to-local scores should be the same.
-    assert_eq(rnkvec_dask, rnkvec_local)
+    assert_eq(rnkvec_dask, rnkvec_local_1)
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:396]:
```suggestion
    assert_eq(rnkvec_dask, rnkvec_dask_local)
```

File: tests/python_package_test/test_dask.py [hunk 8]
```diff
@@ -300,10 +276,12 @@ def test_regressor(output, client, listen_port):
     # Scores should be the same
     if output != 'dataframe':
         assert_eq(s1, s2, atol=.01)
+        assert_eq(s1, s2_local)
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:279]:
```suggestion
        assert_eq(s1, s1_local)
```

File: tests/python_package_test/test_dask.py [hunk 9]
```diff
@@ -300,10 +276,12 @@ def test_regressor(output, client, listen_port):
...
 
     # Predictions should be roughly the same
     assert_eq(y, p1, rtol=1., atol=100.)
     assert_eq(y, p2, rtol=1., atol=50.)
+    assert_eq(p1, p2_local)
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:284]:
```suggestion
    assert_eq(p1, p1_local)
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
I left a few other minor requested changes. Once you do this, I think this will be good to merge.

Really appreciate you taking the time to contribute!

**Review by jameslamb — APPROVED** [review]
Thanks so much! I just added a commit resolving conflicts with some recently-merged stuff on `master`. I'll merge this once it builds.

We hope you'll contribute more to LightGBM in the future!

**Comment by ShrillShrestha** [comment]
Thank you @jameslamb for having your patience with me. I got to learn alot form you just by working on this issue. I'll look forward to contribute more in future. :)

File: tests/python_package_test/test_dask.py [hunk 10]
```diff
@@ -276,12 +276,12 @@ def test_regressor(output, client, listen_port):
     # Scores should be the same
     if output != 'dataframe':
         assert_eq(s1, s2, atol=.01)
-        assert_eq(s1, s2_local)
+        assert_eq(s1, s1_local)
```
Comment (ShrillShrestha) [inline tests/python_package_test/test_dask.py:288]: I saw that one of the tests is failing. I checked the log, and this is the line that is giving an error. Is it because I didn't pass an `atol` key argument?

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:288]: yes I think so. I can see in the logs that the differences are very small, maybe small enough to be due to precision differences

For example, from https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=8740&view=logs&j=c28dceab-947a-5848-c21f-eef3695e5f11&t=fa158246-17e2-53d4-5936-86070edbaacf

> a_original = 0.9555812336527378, b_original = 0.9555698440069504

Please add `atol = 0.003` here for now. That will be enough to catch major errors. We have an open issue (#3835) to come back in the future and investigate how to close these gaps.

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:288]: Note that I already resolved conflicts with `master` on your branch. So you should pull this branch fresh before working on it.

```shell
git branch -D reduce-tests
git fetch origin reduce-tests
git checkout reduce-tests
```

Comment (ShrillShrestha) [reply tests/python_package_test/test_dask.py]: I have made the change. It should work now.

**Review by jameslamb — APPROVED** [review]
looks good, thank you very much for your contribution

**Comment by ShrillShrestha** [comment]
Thanks, @jameslamb for helping me out.
